### PR TITLE
Feat/payout

### PR DESCRIPTION
**Pull request by ymc182**
Takes #673 closer to the final stage

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -2,7 +2,7 @@
 members = [
     "near-sdk",
     "near-sdk-macros",
-    "near-sdk-sim",
+		"near-sdk-sim",
```
Comment (austinabell) [inline Cargo.toml:5]: can you revert this spacing?

File: near-contract-standards/src/non_fungible_token/core/core_impl.rs [hunk 2]
```diff
@@ -426,7 +431,10 @@ impl NonFungibleTokenCore for NonFungibleToken {
         msg: String,
     ) -> PromiseOrValue<bool> {
         assert_one_yocto();
-        require!(env::prepaid_gas() > GAS_FOR_NFT_TRANSFER_CALL, "More gas is required");
+        require!(
+            env::prepaid_gas() > GAS_FOR_NFT_TRANSFER_CALL + GAS_FOR_RESOLVE_TRANSFER,
```
Comment (austinabell) [inline near-contract-standards/src/non_fungible_token/core/core_impl.rs:435]: revert this change please

File: near-contract-standards/src/non_fungible_token/mod.rs [hunk 3]
```diff
@@ -12,6 +12,9 @@ mod macros;
 /// Metadata traits and implementation according to the [NFT enumeration standard](https://nomicon.io/Standards/NonFungibleToken/Metadata.html).
 /// This covers both the contract metadata and the individual token metadata.
 pub mod metadata;
+
+pub mod payout;
```
Comment (austinabell) [inline near-contract-standards/src/non_fungible_token/mod.rs:16]: Module isn't documented here or in `.../payout/mod.rs`,

can you describe this and link to the spec there using `//!` rustdoc comments at the top of the file, please?

File: near-contract-standards/src/non_fungible_token/payout/payout_impl.rs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+use crate::non_fungible_token::NonFungibleToken;
+use near_sdk::assert_one_yocto;
+use near_sdk::{require, AccountId, Balance, IntoStorageKey};
+use std::collections::HashMap;
+impl Royalties {
+    /*   pub fn new(accounts: HashMap<AccountId, u8>, percent: u8) -> Self {
+        let this = Self { accounts, percent };
+        this.validate();
+        this
+    } */
```
Comment (austinabell) [inline near-contract-standards/src/non_fungible_token/payout/payout_impl.rs:13]:
```suggestion
```

File: near-contract-standards/src/non_fungible_token/payout/payout_impl.rs [hunk 5]
```diff
@@ -0,0 +1,80 @@
+use super::NonFungibleTokenPayout;
+use crate::non_fungible_token::core::NonFungibleTokenCore;
+use crate::non_fungible_token::payout::*;
+use crate::non_fungible_token::NonFungibleToken;
+use near_sdk::assert_one_yocto;
+use near_sdk::{require, AccountId, Balance, IntoStorageKey};
+use std::collections::HashMap;
+impl Royalties {
```
Comment (austinabell) [inline near-contract-standards/src/non_fungible_token/payout/payout_impl.rs:8]:
```suggestion
use std::collections::HashMap;

impl Royalties {
```

File: near-contract-standards/src/non_fungible_token/payout/payout_impl.rs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        };
+        let rest = balance - royalty_payment;
+        let owner_payout: u128 = payout.payout.get(owner_id).map_or(0, |x| x.0) + rest;
+        payout.payout.insert(owner_id.clone(), owner_payout.into());
+        payout
+    }
+}
+
+fn apply_percent(percent: u8, int: u128) -> u128 {
+    int * percent as u128 / 100u128
```
Comment (austinabell) [inline near-contract-standards/src/non_fungible_token/payout/payout_impl.rs:56]: Seems like this can overflow, are there any guarantees around this? Probably want to do a checked multiplication with an `expect` to make sure someone doesn't need overflow-checks enabled to catch this

**Review by austinabell — CHANGES_REQUESTED** [review]
Quick scan. @jlogelin can you review that this matches the spec?

I don't have context on this and unclear how it's assumed the royalty percentage would be set/updated. Also didn't check closely if this matches the spec

File: Cargo.toml (hunk 1 above)
Comment (ymc182) [reply Cargo.toml:5]: No Problem

File: near-contract-standards/src/non_fungible_token/payout/mod.rs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+#[serde(crate = "near_sdk::serde")]
+
+pub struct Payout {
+    pub payout: HashMap<AccountId, U128>,
+}
+#[derive(Deserialize, Serialize, BorshDeserialize, BorshSerialize, Default, Debug)]
+#[serde(crate = "near_sdk::serde")]
+pub struct Royalties {
+    key_prefix: Vec<u8>,
+    pub accounts: HashMap<AccountId, u8>,
```
[the remark this answers is not in the document]
Comment (ymc182) [reply near-contract-standards/src/non_fungible_token/payout/mod.rs:18]: No Problem

**Comment by thaibao-incognito** [comment]
How long this feature should be merged into?

**Comment by willemneal** [comment]
@thaibao-incognito, there is no timeline.  If you want to use this today, here is a complete example: https://github.com/TENK-DAO/tenk/blob/main/contracts/tenk/src/payout.rs

**Comment by frol** [comment]
FYI, Payout NFT standard was aligned with the implementation: https://github.com/near/NEPs/pull/276

This PR is almost there, and it seems only the last mile is left, so I hope NEAR community can make this final push here

**Comment by ruseinov** [comment]
@frol I think this can now be closed in favour of https://github.com/near/near-sdk-rs/pull/1077
